torch.package change `Folder` to `Directory` and add doc strings

lobstergrindset opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#56925 torch.package change `Folder` to `Directory` and add doc strings**

Get documentation ready to be public for `Folder` representaiton. Changed `Folder` to `Directory`, added doc strings, and added to the package's .rst file for rendering 

Differential Revision: [D28002145](https://our.internmc.facebook.com/intern/diff/D28002145)

Commit:
torch.package change `Folder` to `Directory` and add doc strings

[ghstack-poisoned]

Commit:
Update on "torch.package change `Folder` to `Directory` and add doc strings"


Differential Revision: [D28002145](https://our.internmc.facebook.com/intern/diff/D28002145)

[ghstack-poisoned]

SplitInfinity commented on torch/package/_file_structure_representation.py:
```diff
@@ -4,27 +4,47 @@
 
-class Folder:
+class Directory:
+    """A file structure representation. Organized as Directory nodes who have lists of
```
```suggestion
    """A file structure representation. Organized as Directory nodes that have lists of
```

SplitInfinity replied:
Also, consider using 
```
:class:`Directory`
```
instead of just `Directory`. This will render in the API reference as a link to the `Directory` class. But maybe it's not useful to have that link because this *is* the `Directory` class; it makes more sense when the docstring refers to other classes.

Lilyjjo replied:
I used the `:class:'Directory'` string in the other public API to enable linking 👍

SplitInfinity commented on torch/package/_file_structure_representation.py:
```diff
@@ -4,27 +4,47 @@
-        self.children: Dict[str, Folder] = {}
+        self.children: Dict[str, Directory] = {}
+
+    def get_dir(self, dirs: List[str]):
```
```suggestion
    def get_dir(self, dirs: List[str]):
```
Consider adding a return type annotation.

SplitInfinity replied:
And a `Returns:` section to the docstring.

SplitInfinity commented on torch/package/_file_structure_representation.py:
```diff
@@ -4,27 +4,47 @@
+            dirs (List[str]): List of directory names that are treated like a path.
+        """
+        # builds path of directories if not yet built, returns last directory
+        if len(dirs) == 0:
```
```suggestion
        if not dirs:
```

suo replied:
chiming in randomly, I think `not dirs` is less clear (introduces local ambiguity—are we checking `None`-ness? `False`-ness?)

SplitInfinity commented on torch/package/_file_structure_representation.py:
```diff
@@ -4,27 +4,47 @@
+            file_path (str): Path of file to add. Last element is added as a file while
+                other paths items are added as directories.
+        """
+        *dirs, file = file_path.split("/")
```
Consider using [`os.sep`](https://docs.python.org/3/library/os.html#os.sep) instead of `"/"`. Also, there's an ominous note in the documentation for `os.sep` that knowing the separator is not sufficient for correct path manipulation and parsing.

SplitInfinity commented on torch/package/_file_structure_representation.py:
```diff
@@ -4,27 +4,47 @@
+        dir.children[file] = Directory(file, False)
 
     def has_file(self, filename: str):
+        """ Returns `True` if a Directory contains the specified file.
```
Can you put this in a `Returns` section? That will get it to render properly in the API reference:

https://github.com/pytorch/pytorch/blob/master/torch/package/package_importer.py#L115-L116

SplitInfinity commented on torch/package/_file_structure_representation.py:
```diff
@@ -75,16 +96,29 @@ def stringify_tree(
     exclude: "GlobPattern" = (),
-) -> Folder:
+) -> Directory:
+    """ Return a Directory file structure representation created from a list of files.
```
Same here, consider moving this to a `Returns` section.

SplitInfinity reviewed (commented):
Nits

Commit:
Update on "torch.package change `Folder` to `Directory` and add doc strings"


Differential Revision: [D28002145](https://our.internmc.facebook.com/intern/diff/D28002145)

[ghstack-poisoned]

Lilyjjo commented:
Since the last review I addressed all the comments and changed the `Directory` object to be public instead of private. I updated the package .rst also

Commit:
Update on "torch.package change `Folder` to `Directory` and add doc strings"


Get documentation ready to be public for `Folder` representaiton. Changed `Folder` to `Directory`, added doc strings, and added to the package's .rst file for rendering 

Differential Revision: [D28002145](https://our.internmc.facebook.com/intern/diff/D28002145)

[ghstack-poisoned]

suo commented on torch/package/file_structure_representation.py:
```diff
@@ -0,0 +1,132 @@
+
+
+class Directory:
+    """A file structure representation. Organized as Directory nodes that have lists of
```
Might want to note that this is returned from ``:meth:`PackageExporter.file_structure` `` and ``:meth:`PackageImporter.file_structure` `` so that folks looking at this class context-free understand how it is used.

Commit:
Update on "torch.package change `Folder` to `Directory` and add doc strings"


Get documentation ready to be public for `Folder` representaiton. Changed `Folder` to `Directory`, added doc strings, and added to the package's .rst file for rendering 

Differential Revision: [D28002145](https://our.internmc.facebook.com/intern/diff/D28002145)

[ghstack-poisoned]
